File: frontend/src/api/llm.ts
import apiClient from './client'

export interface CLIStatus {
  tool: string
  installed: boolean
  version: string | null
  latest_version: string | null
  is_outdated: boolean
  path: string | null
  install_command: string
  update_command: string | null
  platform: string | null
}

export interface LLMProvider {
  id: string
  name: string
  description: string
  configured: boolean
  env_configured: boolean  // Whether API key is set in server .env
  user_configured: boolean // Whether API key is set by user
  is_primary: boolean
  models: string[]
  default_model: string
  selected_model: string
}

export interface LLMConfigResponse {
  preferred_llm: string
  openai_configured: boolean
  anthropic_configured: boolean
  gemini_configured: boolean
  openai_model: string
  anthropic_model: string
  gemini_model: string
  claude_code_status: CLIStatus
  gemini_cli_status: CLIStatus | null
  codex_cli_status: CLIStatus | null
  providers: LLMProvider[]
  runtime: string
}

export interface LLMConfigUpdate {
  provider?: string
  openai_api_key?: string
  anthropic_api_key?: string
  gemini_api_key?: string
  openai_model?: string
  anthropic_model?: string
  gemini_model?: string
}

export interface APIKeyValidationResponse {
  valid: boolean
  error: string | null
  provider: string
}

export interface LLMProviderInfo {
  id: string
  name: string
  description: string
  models: string[]
  default_model: string
  docs_url: string
  has_cli: boolean
}

export interface CLITestResponse {
  success: boolean
  tool: string
  message: string
  output?: string
  provider?: string
  token_usage?: number
  tokens?: number
  auth_status?: 'authenticated' | 'auth_failed' | 'unknown'
}

export interface CLIConnectResponse {
  success: boolean
  message: string
  provider?: string
  env_var?: string
}

export interface LLMTestRequest {
  api_key?: string
  model?: string
}

export interface LLMTestResponse {
  success: boolean
  provider: string
  model: string
  response: string
  token_usage?: number
}

// CLI Native Auth (web-local mode)
export interface CLIAuthStatusResponse {
  authenticated: boolean
  method?: 'oauth' | 'api_key'
  email?: string
  account?: string
  error?: string
}

export interface CLILoginResponse {
  success: boolean
  url?: string
  message?: string
  device_code?: string
}

export interface CLILogoutResponse {
  success: boolean
  message?: string
}

export const llmApi = {
  // Get LLM configuration and CLI status
  getConfig: (userId?: number) =>
    apiClient.get<LLMConfigResponse>('/llm/config', {
      params: userId ? { user_id: userId } : undefined,
    }),

  // Update LLM configuration (API keys and provider)
  updateConfig: (data: LLMConfigUpdate, userId?: number) =>
    apiClient.put<LLMConfigResponse>('/llm/config', data, {
      params: userId ? { user_id: userId } : undefined,
    }),

  // Validate an API key for a provider
  validateKey: (provider: string, apiKey: string) =>
    apiClient.post<APIKeyValidationResponse>(`/llm/validate/${provider}`, { api_key: apiKey }),

  // Get Claude Code CLI status
  getCLIStatus: () =>
    apiClient.get<CLIStatus>('/llm/cli/status'),

  // Get Gemini CLI status
  getGeminiCLIStatus: () =>
    apiClient.get<CLIStatus>('/llm/cli/gemini/status'),

  // Get Codex CLI status
  getCodexCLIStatus: () =>
    apiClient.get<CLIStatus>('/llm/cli/codex/status'),

  // Force refresh CLI status
  refreshCLI: () =>
    apiClient.post<CLIStatus>('/llm/cli/refresh'),

  // List available providers
  getProviders: () =>
    apiClient.get<LLMProviderInfo[]>('/llm/providers'),

  // Test CLI tool (with optional model parameter)
  testCLI: (cliType: 'claude_code' | 'gemini_cli' | 'codex_cli', model?: string) =>
    apiClient.post<CLITestResponse>(`/llm/cli/test/${cliType}`, null, {
      params: model ? { model } : undefined,
    }),

  // Connect CLI tool (save API key to .env)
  connectCLI: (cliType: 'claude_code' | 'gemini_cli' | 'codex_cli', apiKey: string) =>
    apiClient.post<CLIConnectResponse>(`/llm/cli/connect/${cliType}`, { api_key: apiKey }),

  // Disconnect CLI tool (remove API key from .env)
  disconnectCLI: (cliType: 'claude_code' | 'gemini_cli' | 'codex_cli') =>
    apiClient.post<CLIConnectResponse>(`/llm/cli/disconnect/${cliType}`),

  // CLI Native Auth (web-local mode)
  getCLIAuthStatus: (cliType: string) =>
    apiClient.get<CLIAuthStatusResponse>(`/llm/cli/auth/${cliType}`),

  startCLILogin: (cliType: string) =>
    apiClient.post<CLILoginResponse>(`/llm/cli/auth/${cliType}/login`),

  cancelCLILogin: () =>
    apiClient.post(`/llm/cli/auth/cancel`),

  submitAuthCode: (code: string) =>
    apiClient.post<{ success: boolean; message?: string }>(`/llm/cli/auth/submit-code`, { code }),

  cliLogout: (cliType: string) =>
    apiClient.post<CLILogoutResponse>(`/llm/cli/auth/${cliType}/logout`),

  // Test LLM provider
  // api_key: directly test with this key (without saving)
  // use_env: true for Web (use .env API keys), false for Electron (user-configured keys only)
  testProvider: (providerId: string, options?: {
    userId?: number
    useEnv?: boolean
    apiKey?: string
    model?: string
  }) => {
    const { userId, useEnv = true, apiKey, model } = options || {}
    const requestBody: LLMTestRequest | null = apiKey ? { api_key: apiKey, model } : null

    return apiClient.post<LLMTestResponse>(`/llm/test/${providerId}`, requestBody, {
      params: {
        ...(userId ? { user_id: userId } : {}),
        use_env: useEnv,
      },
    })
  },
}

File: api/schemas/llm.py
"""
LLM Configuration Schemas
"""

from pydantic import BaseModel, Field
from typing import Optional, List


class CLIStatus(BaseModel):
    """Status of a CLI tool installation."""

    tool: str = Field(
        ..., description="Tool identifier: 'claude_code', 'gemini_cli', or 'codex_cli'"
    )
    installed: bool = Field(..., description="Whether the CLI is installed")
    version: Optional[str] = Field(None, description="Installed version")
    latest_version: Optional[str] = Field(None, description="Latest available version")
    is_outdated: bool = Field(False, description="Whether an update is available")
    path: Optional[str] = Field(None, description="Path to CLI executable")
    install_command: str = Field(
        "", description="Platform-specific installation command"
    )
    update_command: Optional[str] = Field(
        None, description="Command to update the CLI (if different from install)"
    )
    platform: Optional[str] = Field(None, description="Detected platform")


class LLMProvider(BaseModel):
    """Information about an LLM provider."""

    id: str = Field(
        ..., description="Provider identifier: 'openai', 'anthropic', 'gemini'"
    )
    name: str = Field(..., description="Display name")
    description: str = Field("", description="Provider description")
    configured: bool = Field(
        False, description="Whether API key is configured (user or env)"
    )
    env_configured: bool = Field(
        False, description="Whether API key is set in server .env"
    )
    user_configured: bool = Field(False, description="Whether API key is set by user")
    is_primary: bool = Field(False, description="Whether this is the selected provider")
    models: List[str] = Field(default_factory=list, description="Available models")
    default_model: str = Field("", description="Default model for this provider")
    selected_model: str = Field(
        "", description="Currently selected model for this provider"
    )


class LLMConfigResponse(BaseModel):
    """Response for LLM configuration."""

    preferred_llm: str = Field(..., description="Currently selected LLM provider")
    openai_configured: bool = Field(False, description="Whether OpenAI API key is set")
    anthropic_configured: bool = Field(
        False, description="Whether Anthropic API key is set"
    )
    gemini_configured: bool = Field(False, description="Whether Gemini API key is set")
    openai_model: str = Field("gpt-4.1", description="Selected OpenAI model")
    anthropic_model: str = Field(
        "claude-sonnet-4-6-20260217", description="Selected Anthropic model"
    )
    gemini_model: str = Field("gemini-2.5-flash", description="Selected Gemini model")
    claude_code_status: CLIStatus = Field(..., description="Claude Code CLI status")
    gemini_cli_status: Optional[CLIStatus] = Field(
        None, description="Gemini CLI status"
    )
    codex_cli_status: Optional[CLIStatus] = Field(None, description="Codex CLI status")
    providers: List[LLMProvider] = Field(
        default_factory=list, description="Available providers"
    )
    runtime: str = Field(
        "external",
        description="Runtime environment: 'docker', 'local', 'electron', or 'external'",
    )


class LLMConfigUpdate(BaseModel):
    """Request to update LLM configuration."""

    provider: Optional[str] = Field(None, description="Set preferred provider")
    openai_api_key: Optional[str] = Field(
        None, description="OpenAI API key (will be encrypted)"
    )
    anthropic_api_key: Optional[str] = Field(
        None, description="Anthropic API key (will be encrypted)"
    )
    gemini_api_key: Optional[str] = Field(
        None, description="Gemini API key (will be encrypted)"
    )
    openai_model: Optional[str] = Field(None, description="Selected OpenAI model")
    anthropic_model: Optional[str] = Field(None, description="Selected Anthropic model")
    gemini_model: Optional[str] = Field(None, description="Selected Gemini model")


class APIKeyValidationRequest(BaseModel):
    """Request to validate an API key."""

    api_key: str = Field(..., description="API key to validate")


class APIKeyValidationResponse(BaseModel):
    """Response from API key validation."""

    valid: bool = Field(..., description="Whether the API key is valid")
    error: Optional[str] = Field(None, description="Error message if validation failed")
    provider: str = Field(..., description="Provider that was validated")


class LLMProviderInfo(BaseModel):
    """Information about available LLM providers."""

    id: str
    name: str
    description: str
    models: List[str]
    default_model: str
    docs_url: str
    has_cli: bool = False  # Claude Code and Gemini CLI are supported


class CLITestResponse(BaseModel):
    """Response from CLI test."""

    success: bool = Field(..., description="Whether the test was successful")
    tool: str = Field(..., description="Tool that was tested")
    message: str = Field(..., description="Test result message")
    output: Optional[str] = Field(None, description="CLI output if any")
    model: Optional[str] = Field(None, description="Model used for the test")
    provider: Optional[str] = Field(
        None, description="Mapped provider for usage tracking"
    )
    token_usage: int = Field(default=0, description="Token usage from actual API call")
    auth_status: str = Field(
        default="unknown",
        description="Authentication status: 'authenticated', 'auth_failed', or 'unknown'",
    )


class CLIConnectRequest(BaseModel):
    """Request to connect a CLI tool by providing an API key."""

    api_key: str = Field(..., description="API key for the CLI tool's provider")


class CLIConnectResponse(BaseModel):
    """Response from CLI connect/disconnect."""

    success: bool = Field(..., description="Whether the operation was successful")
    message: str = Field(..., description="Result message")
    provider: Optional[str] = Field(None, description="Mapped provider ID")
    env_var: Optional[str] = Field(
        None, description="Environment variable that was set/unset"
    )


class LLMTestRequest(BaseModel):
    """Request to test an LLM provider with a specific API key."""

    api_key: Optional[str] = Field(
        None, description="API key to test (if not provided, uses stored key)"
    )
    model: Optional[str] = Field(None, description="Model to use for testing")


class StoredAPIKeysResponse(BaseModel):
    """Response containing stored (decrypted) API keys for a user."""

    openai_api_key: Optional[str] = Field(None, description="Stored OpenAI API key")
    anthropic_api_key: Optional[str] = Field(
        None, description="Stored Anthropic API key"
    )
    gemini_api_key: Optional[str] = Field(None, description="Stored Gemini API key")


class LLMTestResponse(BaseModel):
    """Response from LLM provider test."""

    success: bool = Field(..., description="Whether the test was successful")
    provider: str = Field(..., description="Provider that was tested")
    model: str = Field(..., description="Model that was used")
    response: str = Field(..., description="LLM response")
    token_usage: int = Field(default=0, description="Total tokens used")
